model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java

```diff
@@ -495,4 +510,23 @@ protected void configureTransport(GlobalConfigurationBuilder gcb, String nodeNam
         }
     }
 
+    private void registerSystemWideListeners(KeycloakSession session) {
+        KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
+        ClusterProvider cluster = session.getProvider(ClusterProvider.class);
+        cluster.registerListener(REALM_CLEAR_CACHE_EVENTS, (ClusterEvent event) -> {
+            if (event instanceof ClearCacheEvent) {
+                sessionFactory.invalidate(ObjectType._ALL_);
+            }
+        });
+        cluster.registerListener(REALM_INVALIDATION_EVENTS, (ClusterEvent event) -> {
```

Comment:
Few points:
- The REALM_INVALIDATION_EVENTS is currently sent from RealmCacheSession.runInvalidations() and it is sent with "ignoreSender" true. This means that listener is not called on the sender node itself (It is sent on all other cluster nodes and on all other nodes on all datacenters, but not on sender itself). It looks that this is not correct as "sessionFactory.invalidate" needs to be called on the caller node as well?

- I think that in the case of disabled realm cache, the REALM_INVALIDATION_EVENTS won't be sent at all. Hence listener also won't be probably called. For this one, it won't be called on any cluster/cross-dc nodes

Reply:
> The REALM_INVALIDATION_EVENTS is currently sent from RealmCacheSession.runInvalidations() and it is sent with "ignoreSender" true. This means that listener is not called on the sender node itself (It is sent on all other cluster nodes and on all other nodes on all datacenters, but not on sender itself). It looks that this is not correct as "sessionFactory.invalidate" needs to be called on the caller node as well?

This is a great observation. You're right that the `REALM_INVALIDATION_EVENTS` is not received on the sender node, and that is intentional. The REALM_INVALIDATION_EVENTS is specific to `DefaultInfinispanConnectionProviderFactory` or in general to `infinispan` model. The invalidations for the local node are expected to be triggered by the appropriate operation, and be sufficiently fine-grained. In this PR, this is covered only by a detailed `COMPONENT` invalidation upon component config change / removal in https://github.com/keycloak/keycloak/blob/3cac606746cdcc8e1c1a91073f4ca402f143ac62/server-spi-private/src/main/java/org/keycloak/component/AmphibianProviderFactory.java#L66-L87

However I can see that this can be improved since there should be an invalidation repeated at the end of the session, after all the changes are committed. I'll update the code accordingly.

> I think that in the case of disabled realm cache, the REALM_INVALIDATION_EVENTS won't be sent at all. Hence listener also won't be probably called. For this one, it won't be called on any cluster/cross-dc nodes

This is a valid concern, thank you for that. The cluster events will only work with some cluster provider available. Thus the `DefaultComponentFactoryProviderFactory` only performs caching if the messages are available:
https://github.com/keycloak/keycloak/blob/3cac606746cdcc8e1c1a91073f4ca402f143ac62/services/src/main/java/org/keycloak/services/DefaultComponentFactoryProviderFactory.java#L70-L73 and https://github.com/keycloak/keycloak/blob/3cac606746cdcc8e1c1a91073f4ca402f143ac62/services/src/main/java/org/keycloak/services/DefaultComponentFactoryProviderFactory.java#L127-L129

But now I can see that this is insufficient since availability `InfinispanClusterProvider` would not be sufficient. At the same time, binding the caching availability to `CachedRealmProvider` would not be correct as well as in the new store there would be no `CachedRealmProvider`. I'll introduce a property of the `DefaultComponentFactoryProviderFactory` that would disable caching explicitly then, effectively reverting the functionality back to current behaviour.